README.md
https://www.curseforge.com/wow/addons/whitepaws

Only for chinese tbc classic version

1. Help with cat dps skill macro
Can make macro let it auto detect either spell attack skill or shift form
conditions:energy, mana, been rooted, slowed, or strong control like stun or fear.

2. Help with bear form
Can bind auto unroot in skill macros
Can make macro let bear auto choose skill or shift if not enough rage.

3. Auto equip speed up trinket when mounting, equip back when not mounting

4. Tells if been control
 
input /wc for help and settings

只能用于中文燃烧的远征怀旧客户端
1. 猫输出宏
帮助写出输出宏，自动决定打出技能还是变身
考虑的状态有：能量，法力，是不是被定身减速，是不是被晕恐惧之类硬控

2. 熊用的宏
可以把熊的拉怪宏绑定自动解除定身功能
帮助写出宏让熊在不够能量时选择重新变身还是打技能
 

3. 坐骑时自动装加速饰品,不飞时装回

4. 被控时会提示
 
输入 /wc 获得帮助和进行设置
 
先把目录文件放到Interface/Add Ons/WhitePaws里，插件选择上WhitePaws,就可以用说明.txt里面的宏了

猫输出宏和个别熊输出宏需要提前准备好一个[闪光的小珠]，钓鱼商人有卖50铜一个。如果钓鱼技能50以上可以买[夜色虫]，唯一的好处就是字数少宏可以写更多内容


CHANGELOG.txt
--10172022 2.2.0
自动出猫判定增加不在清晰预兆，猛虎不在cd；去除猛虎buff判定
自动出熊判定增加自动解控和清晰预兆变猫
备注：自动出猫和猫自动解控现在只能2选1
更新说明文档

--09232022 2.1.0
添加使用变形金刚-随时变形状时在猛虎，狂暴时阻止自动变身
修复一些文本错误

--09172022 2.0.0
适配喵喵王前夕版本，添加变形金刚-随时变形状，白爪生日快乐！

--05252022 1.8.4
优化延迟和手感

--05082022 1.8.3
优化手感和本地延迟

--04292022 1.8.2
版本更新，一些小优化

--02012022 1.8.1
修复一些小错误

--01312022 1.8.0
猫年快乐！优化手感和能量指示条,能量指示条和计算准确率超过其它插件和WA
优化考虑客户端和网络延迟的猫输出函数dpsp
增加一些设置界面

--01292022 1.7.2
优化手感和能量指示条

--11282021 1.7.1
增加猫输出留蓝功能(详细请看说明.txt),自动吃蓝不仅限于boss战了
优化不同boss控制决策，格鲁尔和玛瑟里顿的控制技能识别
优化能量魔法指示条

--11092021 1.7.0
优化了下一跳能量的时间计算方式，现在更准确
解减速一键打法师功能并入默认输出，如果是场景减速的话如猎人陷阱解不了但是可以不点

--10272021 1.6.7
更新了说明文档，说明描述更加细致了
猫德输出宏添加dpsp和dpsl模式
dpsp:pvp模式，打不到目标自动解定身的基础上，增加了自动解减速，现在可以一键打法师了
dpsl:省蓝模式，不遵循1.5秒原则，每点能量尽量用到(最降低手感和dps)

--10142021 1.6.6
--修复其他形态时怒气值计算问题

--10132021 1.6.5
--默认用交通工具自动解除变形，取消这个设置，修复一些小错误

--10102021 1.6.3
--修复一些小错误

--10092021 1.6.0
--加入一些斜杠命令、飞行点取消变形[@cksky]
--插件改名，白爪cksky助手

--10052021 1.5.4
--马鞭优化

--10012021 1.5.3
--被控通告默认关闭

--10012021 1.5.2
--加入使用公用飞行时判断

--09292021 1.5.1
--加入自动解控判断条件：目标还在近战范围不解，除非是首领boss（比如瓦丝琪女士）

--09262021 1.5.0
--帮助熊熊应对瓦丝琪女士，加上了自动解定身绑普通技能熊熊宏
--更新熊的宏例子

--09242021 1.4.5
--cksky贡献马鞭代码
--优化延迟下输出

--09202021 1.4.3
--被换下饰品现在缓存到本地，下线后可以正确换回
--增加新的命令/wcalert 可以开关被控通告
--Welcome notice

--09172021 1.4	飞行时自动装上飞行护符，不飞时自动换回去，白爪生日快乐！

--09142021 1.3
--加入被控喊话，分类被控，一键宏现在可以自动解定身和变形了，
--熊在被除了硬控之外可以正确吃药了(被晕吃药不变人,冰环能吃药)

--09122021 1.2
--优化很多手感，加入buff和被控的event

--09092021 1.12
--把latency加入到powerspark

--09082021 1.11
--修复小bug

--09072021 1.1
--加入latency判断，提前变身进施法队列，根据延迟预支能量
--优化决策 1.5秒变身能提早下一个技能原则
--嵌入PowerSpark nextTick

-- 收集到的控制类型
-- STUN_MECHANIC 昏迷
-- STUN 沉睡 放逐 瘫痪 被闷棍 恐惧 魅惑
-- ROOT 被定身
-- POSSESS 魅惑
-- FEAR_MECHANIC 恐惧
-- CONFUSE 变形 迷惑
-- SCHOOL_INTERRUPT 打断
-- SILENCE 变形 沉默
-- DISARM 缴械
-- FEAR 惊骇
-- CHARM 魅惑
-- PACIFY 平静


WhitePaws.lua
local addonName, wc = ...

--公共函数

--输出，自动解定身减速，考虑延迟和能量延迟，tbc有用，wlk弃用
function dps(cost, mana)
	if not wc.strongControl and wc.enoughMana(mana) and (wc.needUnroot() or wc.ableShift() and not wc.enoughEnergywithNextTickwithDelay(cost)) then
		SetCVar('autoUnshift', 1)
	else
		SetCVar('autoUnshift', 0)
	end
end

--变身
function shift(r, e, m)
	r = r or 200
	e = e or 200
	if not wc.strongControl and wc.enoughMana(m) and not wc.getShiftGCD() and (wc.needUnroot() or wc.getRage() < r and wc.getEnergy() < e) then
		SetCVar('autoUnshift', 1)
	else
		SetCVar('autoUnshift', 0)
	end
end

--吃红，tbc有用，wlk弃用
--9634巨熊形态
function hppot()
	local u,n = IsUsableSpell(9634)
	if wc.getShiftGCD() or wc.strongControl or not u
		then SetCVar('autoUnshift', 0)
	else
		SetCVar('autoUnshift', 1)
	end
end

--变形金刚-变熊
function bxjgx(e)
	if not wc.strongControl and wc.enoughMana() and not wc.getShiftGCD() and not wc.enoughEnergy(e) and not wc.getBuff(50334) and GetSpellCooldown(50213) ~= 0 then
		SetCVar('autoUnshift', 1)
	else
		SetCVar('autoUnshift', 0)
	end
end

--变形金刚-变猫
function bxjgm(e)
	if not wc.strongControl and wc.enoughMana() and (wc.needUnroot() or (wc.enoughEnergy(e) and not wc.getBuff(50334))) then
		if not wc.getShiftGCD() then
			SetCVar('autoUnshift', 1)
		else
			cancelQueue = 1
			SetCVar('autoUnshift', 0)
		end
	else
		SetCVar('autoUnshift', 0)
	end
end

--结束
function wcEnd()
	if cancelQueue == 1 then
		SpellCancelQueuedSpell()
		cancelQueue = 0
	end
	SetCVar('autoUnshift', 1)
	UIErrorsFrame:Clear()
end


Libs/Flight.lua
local addonName, wc = ...

--触发：变形,移动,BUFF,换过装备,变形,脱战
--上坐骑或飞行自动换饰品
--马鞭: 25653 迅捷飞行符咒: 32481 碎天者之鞭: 32863 棍子上的胡萝卜: 37312
local function changeBoostTrinket(self, event, ...)
    if UnitLevel("player") > 70 then return end
	if InCombatLockdown() then return end
	local mountedTrinket = nil
    	if GetItemCount(32863) > 0 then
        	mountedTrinket = 32863
    	elseif GetItemCount(25653) > 0 then
        	mountedTrinket = 25653
    	elseif GetItemCount(37312) > 0 then
        	mountedTrinket = 37312
    	end
	if (not IsInInstance() or wcIsInInstance) and IsMounted() and not UnitOnTaxi('player') then
		if GetInventoryItemID('player', 13) ~= mountedTrinket and GetInventoryItemID('player', 14) ~= mountedTrinket then
            originTrinket = GetInventoryItemID('player', 14)
			EquipItemByName(mountedTrinket, 14)
		end
	elseif GetInventoryItemID('player', 14) == mountedTrinket then
        if originTrinket ~= nil then
			EquipItemByName(originTrinket, 14)
		end
	end
end

local boostFrame = CreateFrame('Frame')
boostFrame:RegisterEvent('PLAYER_MOUNT_DISPLAY_CHANGED')
boostFrame:RegisterEvent('UPDATE_SHAPESHIFT_FORM')
boostFrame:RegisterEvent('PLAYER_STARTED_MOVING')
boostFrame:RegisterEvent('PLAYER_STOPPED_MOVING')
boostFrame:RegisterEvent('UNIT_AURA')
boostFrame:RegisterEvent('PLAYER_EQUIPMENT_CHANGED')
boostFrame:RegisterEvent('PLAYER_REGEN_ENABLED') --比PLAYER_LEAVE_COMBAT更精确的脱战
boostFrame:RegisterEvent('PLAYER_ENTERING_WORLD')
boostFrame:SetScript('OnEvent', changeBoostTrinket)

--点击飞行点地图时自动取消变形
local Shapeshifted, FlightPointButton

function wc.autoUnshiftonTaxi()
    local texture_str = 'Interface\\TARGETINGFRAME\\UI-StatusBar'
    if InCombatLockdown() then return end
    if not wc.autoUnshiftFrame then
        wc.autoUnshiftFrame = CreateFrame('Button', 'unshiftMacroButton', UIParent, 'SecureActionButtonTemplate')
        wc.autoUnshiftFrame:SetAttribute('type1', 'macro')
        wc.autoUnshiftFrame:SetAttribute("macrotext1",'/cancelform\n/script autoCancelShapeshift()\n/script TakeTaxiNode(FlightPoint)')
        wc.autoUnshiftFrame:SetParent(TaxiRouteMap)
        wc.autoUnshiftFrame:SetAllPoints(TaxiRouteMap)
        wc.autoUnshiftFrame:SetSize(16,16)
        wc.autoUnshiftFrame:SetPoint('TOPLEFT',0,0)
        wc.autoUnshiftFrame:EnableMouse(false)
        wc.autoUnshiftFrame:SetFrameLevel(3)
        wc.autoUnshiftFrame:Show()
        wc.autoUnshiftFrame:SetScript("OnUpdate",function(self,motion)
        	autoCancelShapeshiftForm()
        	FlightPoint = FlightPoint or 0
        	if not FlightPointButton then return end
        	if (Shapeshifted or GetShapeshiftFormID()) and MouseIsOver(FlightPointButton) then
            		TaxiNodeOnButtonEnter(FlightPointButton)
			FlightPointButton:LockHighlight()
            		wc.autoUnshiftFrame:EnableMouse(true)
        	end
        	if (not Shapeshifted and not GetShapeshiftFormID()) or (not MouseIsOver(FlightPointButton)) then
            		wc.autoUnshiftFrame:EnableMouse(false)
			FlightPointButton:UnlockHighlight()
        	end
    	end)
    	wc.autoUnshiftFrame:SetScript("OnLeave",function()
        	wc.autoUnshiftFrame:EnableMouse(false)
        	TaxiNodeOnButtonLeave(FlightPointButton)
		FlightPointButton:UnlockHighlight()
    	end)
	end
end

--月光林地德鲁伊免费飞行点自动下马
local function moongladeAutoDismount(self)
    if GossipFrameNpcNameText:GetText() == "希尔瓦·菲纳雯斯" or GossipFrameNpcNameText:GetText() == "布瑟恩·草风" then
        Dismount()
        SelectGossipOption(1)
    else
        GossipTitleButton_OnClick(self, button)
    end
end

GossipTitleButton1:SetScript("OnClick",moongladeAutoDismount)

local function StewardOfTime()
    if GossipFrameNpcNameText:GetText() == "时间管理者" then
        if not wc.GossipUnshiftFrame then
            wc.GossipUnshiftFrame = CreateFrame('Button', 'unshiftMacroButton2', UIParent, 'SecureActionButtonTemplate')
            wc.GossipUnshiftFrame:SetAttribute('type1', 'macro')
            wc.GossipUnshiftFrame:SetAttribute("macrotext1",'/cancelform\n/script autoCancelShapeshift()\n/script SelectGossipOption(1)')
            wc.GossipUnshiftFrame:SetParent(GossipGreetingScrollChildFrame)
            wc.GossipUnshiftFrame:SetAllPoints(GossipGreetingScrollChildFrame)
            wc.GossipUnshiftFrame:SetSize(16,16)
            wc.GossipUnshiftFrame:SetPoint('TOPLEFT',0,0)
            wc.GossipUnshiftFrame:EnableMouse(false)
            wc.GossipUnshiftFrame:SetFrameLevel(6)
            wc.GossipUnshiftFrame:Show()
            wc.GossipUnshiftFrame:SetScript("OnUpdate",function(self,motion)
                autoCancelShapeshiftForm()
                if (Shapeshifted or GetShapeshiftFormID()) and MouseIsOver(GossipTitleButton1) then
                        GossipTitleButton1:LockHighlight()
                        wc.GossipUnshiftFrame:EnableMouse(true)
                end
                if (not Shapeshifted and not GetShapeshiftFormID()) or (not MouseIsOver(GossipTitleButton1)) then
                        wc.GossipUnshiftFrame:EnableMouse(false)
                        GossipTitleButton1:UnlockHighlight()
                end
            end)
            wc.GossipUnshiftFrame:SetScript("OnLeave",function()
                wc.GossipUnshiftFrame:EnableMouse(false)
                GossipTitleButton1:UnlockHighlight()
            end)
        end
    end
end

local StewardOfTimeFrame = CreateFrame("frame")
StewardOfTimeFrame:RegisterEvent("GOSSIP_SHOW")
StewardOfTimeFrame:SetScript("OnEvent",StewardOfTime)

--解除德鲁伊变形
function autoCancelShapeshiftForm()
    if InCombatLockdown() then return end
	local i = 1
	Shapeshifted = false
	while UnitBuff('player', i) do
		local spellId = select(10, UnitBuff('player', i))
    	if spellId == 16591 or spellId == 6405 then
    		Shapeshifted = true
    		break
    	end
		i = i + 1
	end
    if NumTaxiNodes() == 0 then return end
    if Shapeshifted or GetShapeshiftFormID() then
        local num = NumTaxiNodes() or 17
        for i = 1, num, 1 do
            FlightPoint = 0
            local name = TaxiNodeName(i) or "暴风城，艾尔文森林"
            if name == GameTooltipTextLeft1:GetText() then
                FlightPoint = i
                FlightPointButton = _G["TaxiButton"..i]
                break
            end
        end
    end
end

--解除诺格弗格药剂（骷髅）和熊怪形态的变形
--诺格弗格药剂（骷髅）:16591  熊怪形态:6405
function autoCancelShapeshift()
    if InCombatLockdown() or NumTaxiNodes() == 0 then
    else
        local i = 1
		while UnitBuff('player', i) do
			local spellId = select(10, UnitBuff('player', i))
			if spellId == 16591 or spellId == 6405 then
                CancelUnitBuff("player", i)
			end
			i = i + 1
		end
    end
end

--移动速度小框体
function wc.showSpeed()
    if not speedFrame then
        speedFrame = CreateFrame('Frame','MiniMapSpeedFrame', nil, 'ThinGoldEdgeTemplate')
        speedFrame:SetParent(MiniMap)
        speedFrame:SetPoint('TOPRIGHT', 0, -150)
        speedFrame:SetFrameStrata('HIGH')
        speedFrame:SetFrameLevel(9)
        speedFrame:SetMovable(true)
        speedFrame.fs = speedFrame:CreateFontString('MinimapLayerFrameFS', 'ARTWORK')
        speedFrame.fs:SetPoint('CENTER', 0, 0)
        speedFrame.fs:SetFont('Fonts\\ARHei.ttf', 10)
        speedFrame:SetWidth(46)
        speedFrame:SetHeight(17)
        speedFrame:SetScript('OnUpdate',function()
            local playerCurrentSpeed = string.format('%d%%', GetUnitSpeed('player') / 7 * 100)
            speedFrame.fs:SetText(playerCurrentSpeed)
        end)
    end
    if wcSpeed == true then
        speedFrame:Show()
    else
        speedFrame:Hide()
    end
end

local showSpeedFrame = CreateFrame('frame')
showSpeedFrame:RegisterEvent('PLAYER_LOGIN')
showSpeedFrame:RegisterEvent('ADDON_LOADED')
showSpeedFrame:RegisterEvent('PLAYER_ENTERING_WORLD')
showSpeedFrame:SetScript('OnEvent',wc.showSpeed)


Libs/Options.lua
local addonName, wc = ...

local function WhitePaws_Help()
	SELECTED_CHAT_FRAME:AddMessage((wcAlert and '[|cff00ff00开|r]' or '[|cffff0000关|r]')..'/wc alert 开关被控通告功能',255,255,0)
	SELECTED_CHAT_FRAME:AddMessage((wcIsInInstance and '[|cff00ff00开|r]' or '[|cffff0000关|r]')..'/wc bg 开关副本/战场内自动马鞭功能',255,255,0)
	SELECTED_CHAT_FRAME:AddMessage((wcSpeed and '[|cff00ff00开|r]' or '[|cffff0000关|r]')..'/wc speed 开关小地图右下方移动速度显示框体',255,255,0)
	SELECTED_CHAT_FRAME:AddMessage('/wc 查看命令帮助',255,255,0)
end

local function WhitePaws_Command(arg)
	arg1, arg2 = strsplit(" ", strlower(arg))
	if arg1 == 'alert' then
		wcAlert = not wcAlert
		WhitePaws_Help()
	elseif arg1 == 'bg' then
		wcIsInInstance = not wcIsInInstance
		WhitePaws_Help()
	elseif arg1 == 'speed' then
		wcSpeed = not wcSpeed
		wc.showSpeed()
		WhitePaws_Help()
	elseif arg1 == 'end' then
		pcall(wcEnd)
	elseif arg1 == 'bxjgx' then
		pcall(bxjgx, tonumber(arg2))
	elseif arg1 == 'bxjgm' then
		pcall(bxjgm, tonumber(arg2))
	else
		WhitePaws_Help()
	end
end

SlashCmdList['WHITEPAWS'] = WhitePaws_Command
SLASH_WHITEPAWS1 = '/whitepaws'
SLASH_WHITEPAWS2 = '/wc'

local function wcInit()
	wcAlert = wcAlert or false
	wcIsInInstance = wcIsInInstance or false
	if wcSpeed == nil then wcSpeed = true end
	local title = select(2, GetAddOnInfo('whitepaws'))
	SELECTED_CHAT_FRAME:AddMessage('---------------------')
	SELECTED_CHAT_FRAME:AddMessage('欢迎使用'..title,255,255,0)
	SELECTED_CHAT_FRAME:AddMessage('---------------------')
	WhitePaws_Help()
	wc.autoUnshiftonTaxi()
	wc.autoUnshiftFrame:Show()
	wc.autoUnshiftFrame:EnableMouse(false)
	cancelQueue = 0;
end

local initFrame = CreateFrame('Frame')

initFrame:RegisterEvent('PLAYER_LOGIN')
initFrame:SetScript('OnEvent', wcInit)

local panel = CreateFrame("FRAME")
panel.name = "WhitePaws"
InterfaceOptions_AddCategory(panel)

local alertButton = CreateFrame("Button", "alertButton_GlobalName", panel, "UIPanelButtonTemplate")
alertButton:SetPoint("TOPLEFT", 100, -30)
alertButton:SetSize(120 ,22)
alertButton_GlobalNameText:SetText("开关被控通报")
alertButton.tooltip = "开关被控通报功能"
alertButton:SetScript("OnClick", 
  function()
    --do stuff
    WhitePaws_Command('alert')
  end
)

local bgButton = CreateFrame("Button", "bgButton_GlobalName", panel, "UIPanelButtonTemplate")
bgButton:SetPoint("TOPLEFT", 100, -60)
bgButton:SetSize(120 ,22)
bgButton_GlobalNameText:SetText("副本/战场马鞭")
bgButton.tooltip = "开关副本/战场内自动马鞭"
bgButton:SetScript("OnClick", 
  function()
    --do stuff
    WhitePaws_Command('bg')
  end
)

local speedButton = CreateFrame("Button", "speedButton_GlobalName", panel, "UIPanelButtonTemplate")
speedButton:SetPoint("TOPLEFT", 100, -90)
speedButton:SetSize(120 ,22)
speedButton_GlobalNameText:SetText("移动速度显示")
speedButton.tooltip = "开关小地图右下方移动速度显示框体"
speedButton:SetScript("OnClick", 
  function()
    --do stuff
    WhitePaws_Command('speed')
  end
)

Libs/PowerSpark.lua
local addonName, wc = ...

local class = select(2, UnitClass('player'))
local notNormalTick = GetTime()
local frame = CreateFrame('Frame')
for _, item in pairs({
	'PLAYER_ENTERING_WORLD',
	'UNIT_AURA',
	'UNIT_SPELLCAST_SUCCEEDED',
	'UNIT_POWER_FREQUENT',
	'COMBAT_LOG_EVENT_UNFILTERED'
}) do
	frame:RegisterEvent(item, 'player')
end
frame:SetScript('OnEvent', function(self, event, ...)
	if event == 'PLAYER_ENTERING_WORLD' then
		function self.cure(key)
			local type = UnitPowerType('player')
			if key == 'druid' then type = 0 end
			return UnitPower('player', type), type
		end
		function self.rest(key, event, unit, powerType)
			local cure, type = self.cure(key)
			if event == 'COMBAT_LOG_EVENT_UNFILTERED' then --小德变身
				local timestamp, subevent, _, sourceGUID, sourceName, sourceFlags, sourceRaidFlags, destGUID, destName, destFlags, destRaidFlags = CombatLogGetCurrentEventInfo()
				if subevent == 'SPELL_ENERGIZE' and sourceName ==  GetUnitName('player') then
					notNormalTick = GetTime()
					if GetTime() - self[key].timer < 0.02 then
						self[key].timer = self[key].last
					end
				end
			elseif event == 'UNIT_POWER_FREQUENT' and unit == 'player' then -- 能量/法力更新
				if cure > self[key].cure and GetTime() - notNormalTick >= 0.02 then
					if GetTime() - self[key].timer >= 0.02 then
						self[key].last = self[key].timer
					end
					self[key].timer = GetTime()
					PowerSparkDB[key].timer = self[key].timer
				end
				self[key].cure = cure
			elseif event == 'UNIT_SPELLCAST_SUCCEEDED' and unit == 'player' then -- 施法成功
				if cure < self[key].cure and type == 0 then self[key].wait = GetTime() + 5 end -- 5秒回蓝
				self[key].cure = cure
			end
		end
		function self.init(parent, key) --初始化
			if not parent or self[key] then return end
			if not PowerSparkDB then PowerSparkDB = {} end
			if not PowerSparkDB[key] then PowerSparkDB[key] = {} end
			local power = CreateFrame('StatusBar', nil, parent)
			power:SetWidth(parent.width or parent:GetWidth())
			power:SetHeight(parent.height or parent:GetHeight())
			power:SetPoint('CENTER')
			power.spark = power:CreateTexture(nil, 'OVERLAY')
			power.spark:SetTexture('Interface\\CastingBar\\UI-CastingBar-Spark')
			power.spark:SetWidth(32)
			power.spark:SetHeight(32)
			power.spark:SetBlendMode('ADD')
			power.spark:SetAlpha(0)
			power.cure = self.cure(key)
			power.rate = GetTime()
			power.timer = PowerSparkDB[key].timer or GetTime()
			power.last = power.timer
			power.interval = 2
			power.key = key
			power.parent = parent
			function power.hide(key)
				local cure, type = self.cure(key)
				return UnitIsDeadOrGhost('player') or key == 'default' and type == 1 or type == 0 and cure >= UnitPowerMax('player', 0) or type == 3 and cure >= UnitPowerMax('player') and not IsStealthed() and (not UnitCanAttack('player', 'target') or UnitIsDeadOrGhost('target')) --角色死亡/怒气/满蓝/满能量且不潜行且目标不可攻击
			end
			power:HookScript('OnUpdate', function(self)
				local now = GetTime()
				if now < self.rate then return end
				self.rate = now + 0.02 --刷新率
				if self.hide(self.key) then
					self.spark:SetAlpha(0)
				elseif self.wait and self.wait > now and (UnitPowerType('player') == 0 or self.key == 'druid') then --5秒等待回蓝
					self.spark:SetAlpha(1)
					self.spark:SetPoint('CENTER', self, 'LEFT', self:GetWidth() * (self.wait - now) / 5, 0)
				elseif self.timer then
					self.spark:SetAlpha(1)
					self.spark:SetPoint('CENTER', self, 'LEFT', self:GetWidth() * (mod(now - self.timer, self.interval) / self.interval), 0)
				end
			end)
			self[key] = power
		end
		if SUFUnitplayer and SUFUnitplayer.powerBar then -- 兼容 SUF
			self.init(SUFUnitplayer.powerBar, 'default')
			if class == 'DRUID' and SUFUnitplayer.druidBar then self.init(SUFUnitplayer.druidBar, 'druid') end
		elseif ElvUF_Player and ElvUF_Player.Power then	-- 兼容 ElvUI
			self.init(ElvUF_Player.Power, 'default')
		elseif StatusBars2_playerPowerBar then	-- 兼容 Statusbars2
			self.init(StatusBars2_playerPowerBar, 'default')
		else
			self.init(PlayerFrameManaBar, 'default')
		end
		if class == 'DRUID' and DruidBarFrame then
			DruidBarFrame.width = DruidBarKey.width - 4
			DruidBarFrame.health = DruidBarKey.height - 4
			self.init(DruidBarFrame, 'druid')
		end
		if class == 'DRUID' and BC_DruidBar then
			BC_DruidBar.width = BC_DruidBar.Mana:GetWidth() - 4
			BC_DruidBar.health = BC_DruidBar.Mana:GetHeight() - 4
			self.init(BC_DruidBar, 'druid')
		end
	elseif event == 'UNIT_AURA' and class == 'ROGUE' then
		self.interval = 2
		local i = 1
		while UnitBuff('player', i) do
			if select(10, UnitBuff('player', i)) == 13750 then --开了冲动
				self.interval = 1
				break
			end
			i = i + 1
		end
	elseif event == 'UNIT_AURA' and class == 'DRUID' then
		local i = 1
		wc.clearcasting = false
		while UnitBuff('player', i) do
			if select(10, UnitBuff('player', i)) == 16870 then --节能施法
				wc.clearcasting = true
				break
			end
			i = i + 1
		end
	elseif self.rest then
		self.rest('default', event, ...)
		if self.druid then self.rest('druid', event, ...) end
	end
end)


Libs/wcCore.lua
local addonName, wc = ...

--判断控制
local bossControl = false

local function GetControls(self, event, unit, ...)
	if event == 'PLAYER_CONTROL_LOST' then
		if wcAlert then
			SELECTED_CHAT_FRAME:AddMessage('哟失去控制了')
		end
		return
	end
	if event == 'PLAYER_CONTROL_GAINED' then
		if wcAlert then
			SELECTED_CHAT_FRAME:AddMessage('哟又能控制了')
		end
		return
	end
	if event == 'UNIT_AURA' then
		if unit == 'player' then
			if wc.getDebuff(33652, 36449) then
				if not bossControl then
					wc.strongControl = true
					bossControl = true
					if wcAlert then
						SendChatMessage('被BOSS强控了，啥也做不了，大家都一样要忍，所以问题不大', 'EMOTE')
					end
				end
			elseif bossControl then
				wc.strongControl = false
				bossControl = false
				if wcAlert then
					SendChatMessage('被BOSS的强控结束了，你长吁了一口气，并说道问题不大', 'EMOTE')
				end
			end
		end
		return
	end
	wc.strongControl = false
	wc.rooted = false
	local eventIndex = C_LossOfControl.GetActiveLossOfControlDataCount()
	while (eventIndex > 0) do
		local locType = C_LossOfControl.GetActiveLossOfControlData(eventIndex).locType
		if locType == 'ROOT' or locType == 'CONFUSE' and C_LossOfControl.GetActiveLossOfControlData(eventIndex).displayText == '变形' then
			wc.rooted = true
			if wcAlert then
				SendChatMessage('['..C_LossOfControl.GetActiveLossOfControlData(eventIndex).displayText..'],还剩'..(math.floor(C_LossOfControl.GetActiveLossOfControlData(eventIndex).timeRemaining  * 10 + 0.5) / 10)..'秒,类型是'..C_LossOfControl.GetActiveLossOfControlData(eventIndex).locType..',能用变形解，所以问题不大', 'EMOTE')
			end
		elseif locType == 'STUN_MECHANIC' or
			locType == 'STUN' or
			locType == 'POSSESS' or
			locType == 'FEAR_MECHANIC' or
			locType == 'CONFUSE' or
			locType == 'FEAR' or
			locType == 'CHARM' then
			wc.strongControl = true
			if wcAlert then
				SendChatMessage('['..C_LossOfControl.GetActiveLossOfControlData(eventIndex).displayText..'],还剩'..(math.floor(C_LossOfControl.GetActiveLossOfControlData(eventIndex).timeRemaining  * 10 + 0.5) / 10)..'秒,类型是'..C_LossOfControl.GetActiveLossOfControlData(eventIndex).locType..',不能用变形解,但是可以忍过去,所以问题不大', 'EMOTE')
			end
		elseif locType == 'SCHOOL_INTERRUPT' or
			locType == 'SILENCE' or
			locType == 'DISARM' or
			locType == 'PACIFY' then
				if wcAlert then
					SendChatMessage('['..C_LossOfControl.GetActiveLossOfControlData(eventIndex).displayText..'],还剩'..(math.floor(C_LossOfControl.GetActiveLossOfControlData(eventIndex).timeRemaining  * 10 + 0.5) / 10)..'秒,类型是'..C_LossOfControl.GetActiveLossOfControlData(eventIndex).locType..',不能用变形解,但是不是硬控,所以问题不大', 'EMOTE')
				end
		else
			if wcAlert then
				SendChatMessage('['..C_LossOfControl.GetActiveLossOfControlData(eventIndex).displayText..'],还剩'..(math.floor(C_LossOfControl.GetActiveLossOfControlData(eventIndex).timeRemaining  * 10 + 0.5) / 10)..'秒,类型是'..C_LossOfControl.GetActiveLossOfControlData(eventIndex).locType..',新的控制技能？', 'EMOTE')
			end
		end
		eventIndex = eventIndex - 1
	end
end

local controlFrame = CreateFrame('Frame')

controlFrame:RegisterEvent('LOSS_OF_CONTROL_UPDATE') -- the player current target recently gained or lost an control
controlFrame:RegisterEvent('PLAYER_CONTROL_LOST') -- the player current target recently gained or lost an control
controlFrame:RegisterEvent('PLAYER_CONTROL_GAINED') -- the player current target recently gained or lost an control
controlFrame:RegisterEvent('UNIT_AURA') -- the player current target recently gained or lost an control
controlFrame:SetScript('OnEvent', GetControls)

function wc.getLatency()
	return select(4, GetNetStats()) / 1000 + 0.025
end

--变形条件
function wc.getShiftGCD()
	return GetSpellCooldown(768) ~= 0
end

function wc.ableShift()
	local cd = 0
	if wc.getShiftGCD() then
		cd = select(2, GetSpellCooldown(768)) + GetSpellCooldown(768) - GetTime()
	end
	return cd <= wc.getLatency()
end

function wc.getRage()
	if GetShapeshiftFormID() ~= 5 and GetShapeshiftFormID() ~= 8  then
		return 0
	end
	return UnitPower('player', 1)
end

function wc.getEnergy()
	return UnitPower('player', 3)
end

--狂暴50334 激怒5229 猛虎9846 50212 50213
function wc.getBuff(...)
	local buffs = {}, i, v
	for i, v in ipairs{...} do
		buffs[v] = true;
	end
	i = 1
	while UnitBuff('player', i) do
		if buffs[select(1, UnitBuff('player', i))] or buffs[select(10, UnitBuff('player', i))] then
			return true
		end
		i = i + 1
	end
	return false
end

function wc.getDebuff(...)
	local debuffs = {}, i, v
	for i, v in ipairs{...} do
		debuffs[v] = true;
	end
	i = 1
	while UnitDebuff('player', i) do
		if debuffs[select(1, UnitDebuff('player', i))] or debuffs[select(10, UnitDebuff('player', i))] then
			return true
		end
		i = i + 1
	end
	return false
end

function wc.enoughMana(cost)
	if cost == nil then
		cost = GetSpellPowerCost(768)[1].cost
	end
	return UnitPower('player', 0) >= cost
end

function wc.enoughEnergy(cost)
	if wc.getBuff(16870) then
		return true
	end
	return wc.getEnergy() >= cost
end

function wc.needUnroot()
	--打得着并且不是瓦斯琪纠缠
	if IsSpellInRange('爪击', 'target') == 1 and not wc.getDebuff(38316) then
		return false
	--定身或者减速了
	elseif wc.rooted then return true
	elseif select(2, GetUnitSpeed('player')) < 7 and not IsStealthed() then return true
	else return false end
	--todo dazed 眩晕 震荡射击
end
